### PR TITLE
Add IPPrefix Presto type

### PR DESCRIPTION
**Pull request by mohsaka**
Written with similar comments addressed from [#10650](https://github.com/facebookincubator/velox/pull/10596)
Part of https://github.com/facebookincubator/velox/pull/10538

Some comments.
1. IPPREFIX uses VARBINARY type, therefore it should not be subject to the same issue as IPADDRESS https://github.com/facebookincubator/velox/issues/10338.
2. The cast to and from functions for IPADDRESS <-> IPPREFIX are both in the IPADDRESS and IPPREFIX. This is not necessary since castTo gets priority. However isSupported* is necessary for the check. I could remove the castFroms and just return in the castFrom function since it will never be called. 
3. Limited fuzzer will be added later. 
4. `auto const vec = splitIpSlashCidr(ipAddressString);` is not outside the switch statement for efficiency. Don't want to re-parse if not necessary. Folly does the same but in a different way which seemed a little messy.
https://github.com/facebook/folly/blob/556e17ac7b1f6cda68de789d5b06e79b2d3b990f/folly/IPAddress.cpp#L84C24-L84C37

### CODE REVIEW

File: velox/docs/develop/types.rst [hunk 1]
```diff
@@ -155,6 +156,18 @@ used in IPADDRESS/IPPREFIX related functions. This type can be used to
...
 IPPREFIX networks.
 
+IPPREFIX represents an IPV6 or IPV4 formatted IPV6 address along with a one byte
+prefix length. Its physical type is VARBINARY but has a fixed length of 17 bytes.
+The format that the address is stored in is defined as part of `(RFC 4291#section-2.5.5.2) <https://datatracker.ietf.org/doc/html/rfc4291.html#section-2.5.5.2>`_
+The prefix length is stored in the last byte of the VARBINARY array.
+The IP address stored is the canonical(smallest) IP address in the
+subnet range. This type can be used in IP subnet functions.
+
+Example::
```
Comment (czentgr) [inline velox/docs/develop/types.rst:166]: Typo, double ::?

File: velox/docs/develop/types.rst [hunk 2]
```diff
@@ -155,6 +156,18 @@ used in IPADDRESS/IPPREFIX related functions. This type can be used to
 create IPPREFIX networks as well as to check IPADDRESS validity within
 IPPREFIX networks.
 
+IPPREFIX represents an IPV6 or IPV4 formatted IPV6 address along with a one byte
+prefix length. Its physical type is VARBINARY but has a fixed length of 17 bytes.
+The format that the address is stored in is defined as part of `(RFC 4291#section-2.5.5.2) <https://datatracker.ietf.org/doc/html/rfc4291.html#section-2.5.5.2>`_
```
Comment (czentgr) [inline velox/docs/develop/types.rst:161]: Missing period after the sentence.

File: velox/docs/develop/types.rst [hunk 3]
```diff
@@ -155,6 +156,18 @@ used in IPADDRESS/IPPREFIX related functions. This type can be used to
...
+The prefix length is stored in the last byte of the VARBINARY array.
+The IP address stored is the canonical(smallest) IP address in the
+subnet range. This type can be used in IP subnet functions.
+
+Example::
+
+   IPPREFIX 'FFFF:FFFF:FFFF:FFFF:FFFF:FFFF:FFFF:FFFF/32' -- IPPREFIX 'FFFF:FFFF:0000:0000:0000:0000:0000:0000/32'
```
Comment (czentgr) [inline velox/docs/develop/types.rst:168]: Is the left hand side the original data input and the right hand side representing what is actually stored based on "The IP address stored is the canonical(smallest) IP address in the
subnet range? Perhaps the example could be a bit more clear on that?

File: velox/docs/functions/presto/conversion.rst [hunk 4]
```diff
@@ -289,6 +304,24 @@ supported conversions to/from JSON are listed in :doc:`json`.
...
+     - Y
+   * - ipprefix
+     - 
+     - 
+     - 
+     - 
+     - 
+     - 
+     - 
+     - Y
```
Comment (czentgr) [inline velox/docs/functions/presto/conversion.rst:316]: If I count this right, this means conversion to VARCHAR is allowed but not VARBINARY? Wouldn't we also allow VARBINARY just like for IPADDRESS?
Or is that not supported by Presto?

File: velox/docs/functions/presto/conversion.rst [hunk 5]
```diff
@@ -289,6 +304,24 @@ supported conversions to/from JSON are listed in :doc:`json`.
...
+     - 
+     -
+     -
+     -
+     -
+     - 
+     -
```
Comment (czentgr) [inline velox/docs/functions/presto/conversion.rst:323]: This is IPADDRESS in the table and we also allow this based on the tests.
The self cast is also working. That one seems to be missing for IPADDRESS (line 306).

File: velox/functions/prestosql/types/IPAddressType.cpp [hunk 6]
```diff
@@ -17,11 +17,15 @@
 #include "velox/functions/prestosql/types/IPAddressType.h"
 #include <folly/IPAddress.h>
 #include "velox/expression/CastExpr.h"
+#include "velox/functions/prestosql/types/IPPrefixType.h"
 
 static constexpr int kIPV4AddressBytes = 4;
 static constexpr int kIPV4ToV6FFIndex = 10;
 static constexpr int kIPV4ToV6Index = 12;
 static constexpr int kIPAddressBytes = 16;
+static constexpr int kIPPrefixBytes = 17;
```
Comment (czentgr) [inline velox/functions/prestosql/types/IPAddressType.cpp:26]: I see multiple definitions in the cpp files for this and for kIPAddressBytes. Shouldn't the common ones be defined in the header file? Perhaps a common header that just defines these consts for IPADDRESS/IPPREFIX?

File: velox/functions/prestosql/types/IPPrefixType.cpp [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+
+      memcpy(&addrBytes, &ipAddress, kIPAddressBytes);
+      std::reverse(addrBytes.begin(), addrBytes.end());
+      memcpy(result.data(), &addrBytes, kIPAddressBytes);
+
+      folly::IPAddressV6 v6Addr(addrBytes);
+      if (v6Addr.isIPv4Mapped()) {
+        result.data()[kIPAddressBytes] = kIPV4Bits;
+      } else {
+        result.data()[kIPAddressBytes] = (unsigned char)kIPV6Bits;
```
Comment (czentgr) [inline velox/functions/prestosql/types/IPPrefixType.cpp:282]: Is this cast needed? Both `kIPV4Bits` and `kIPV6Bits` so not sure what the difference is here.

**Review by czentgr — COMMENTED** [review]
Thank you!

File: velox/docs/functions/presto/conversion.rst (hunk 4 above)
Comment (mohsaka) [reply velox/docs/functions/presto/conversion.rst:316]: It's currently not supported in presto.

File: velox/docs/develop/types.rst (hunk 2 above)
Comment (mohsaka) [reply velox/docs/develop/types.rst:161]: Fixed

File: velox/docs/develop/types.rst (hunk 1 above)
Comment (mohsaka) [reply velox/docs/develop/types.rst:166]: Fixed

File: velox/functions/prestosql/types/IPAddressType.cpp (hunk 6 above)
Comment (mohsaka) [reply velox/functions/prestosql/types/IPAddressType.cpp:26]: Moved to IPAddressType.h since that should be the base class, as IPPREFIX contains an IPADDRESS.

File: velox/docs/develop/types.rst (hunk 3 above)
Comment (mohsaka) [reply velox/docs/develop/types.rst:173]: Added explanation in example section.

File: velox/functions/prestosql/types/IPPrefixType.cpp (hunk 7 above)
Comment (mohsaka) [reply velox/functions/prestosql/types/IPPrefixType.cpp:277]: Update: Re-added unsigned char. Mac-os build check failed due to conversion from 128 -> -128.

File: velox/functions/prestosql/tests/IPPrefixCastTest.cpp [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+#include "velox/functions/prestosql/tests/utils/FunctionBaseTest.h"
+
+namespace facebook::velox::functions::prestosql {
+
+namespace {
+
+class IPPrefixCastTest : public functions::test::FunctionBaseTest {
+ protected:
+  std::optional<std::string> castToVarchar(
+      const std::optional<std::string> input) {
```
Comment (czentgr) [inline velox/functions/prestosql/tests/IPPrefixCastTest.cpp:27]: Is it possible to use `const std::optional<std::string>&` instead of copying the input?

File: velox/functions/prestosql/tests/IPPrefixCastTest.cpp [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    auto result = evaluateOnce<std::string>(
+        "cast(cast(c0 as ipprefix) as varchar)", input);
+    return result;
+  }
+
+  std::optional<std::string> castFromIPAddress(
+      const std::optional<std::string> input) {
```
Comment (czentgr) [inline velox/functions/prestosql/tests/IPPrefixCastTest.cpp:34]: const &

File: velox/functions/prestosql/tests/IPPrefixCastTest.cpp [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    auto result = evaluateOnce<std::string>(
+        "cast(cast(cast(c0 as ipaddress) as ipprefix) as varchar)", input);
+    return result;
+  }
+
+  std::optional<int128_t> castToIPAddress(
+      const std::optional<std::string> input) {
```
Comment (czentgr) [inline velox/functions/prestosql/tests/IPPrefixCastTest.cpp:41]: const &

File: velox/functions/prestosql/tests/IPPrefixCastTest.cpp [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+    auto result = evaluateOnce<int128_t>(
+        "cast(cast(c0 as ipprefix) as ipaddress)", input);
+    return result;
+  }
+};
+
+int128_t stringToInt128(std::string value) {
```
Comment (czentgr) [inline velox/functions/prestosql/tests/IPPrefixCastTest.cpp:48]: const &

Comment (mohsaka) [reply velox/functions/prestosql/tests/IPPrefixCastTest.cpp]: Done

File: velox/functions/prestosql/tests/IPPrefixCastTest.cpp (hunk 10 above)
Comment (mohsaka) [reply velox/functions/prestosql/tests/IPPrefixCastTest.cpp]: Done

File: velox/functions/prestosql/tests/IPPrefixCastTest.cpp (hunk 9 above)
Comment (mohsaka) [reply velox/functions/prestosql/tests/IPPrefixCastTest.cpp]: Done

File: velox/functions/prestosql/tests/IPPrefixCastTest.cpp (hunk 8 above)
Comment (mohsaka) [reply velox/functions/prestosql/tests/IPPrefixCastTest.cpp]: Done

**Review by czentgr — COMMENTED** [review]
Thanks!
@mbasmanova when you get a chance can you please take a look?

**Review by mbasmanova — COMMENTED** [review]
@mohsaka Thank you for adding this. Looks good, but I haven't read every line. I assume @czentgr has reviewed carefully. Let me know if that's not the case.

File: velox/functions/prestosql/types/IPAddressType.cpp [hunk 12]
```diff
@@ -206,6 +214,57 @@ class IPAddressCastOperator : public exec::CastOperator {
...
+
+      memcpy(&addrBytes, &ipAddress, kIPAddressBytes);
+      std::reverse(addrBytes.begin(), addrBytes.end());
+      memcpy(result.data(), &addrBytes, kIPAddressBytes);
+
+      folly::IPAddressV6 v6Addr(addrBytes);
+      if (v6Addr.isIPv4Mapped()) {
+        result.data()[kIPAddressBytes] = kIPV4Bits;
+      } else {
+        result.data()[kIPAddressBytes] = (unsigned char)kIPV6Bits;
```
Comment (czentgr) [inline velox/functions/prestosql/types/IPAddressType.cpp:262]: We can get rid of the cast as well.

**Review by czentgr — COMMENTED** [review]
@mbasmanova Thank you for your review. But I was not nearly careful enough given the issues you found that are real. I will have to re-review after all this is fixed.

File: velox/functions/prestosql/types/IPPrefixType.cpp [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+      folly::ByteArray16 addrBytes;
+      std::string s;
+
+      memcpy(&addrBytes, intAddr.data(), kIPAddressBytes);
+      folly::IPAddressV6 v6Addr(addrBytes);
+
+      exec::StringWriter<false> result(flatResult, row);
+      if (v6Addr.isIPv4Mapped()) {
+        result.append(
+            v6Addr.createIPv4().str() + "/" +
```
[the remark this answers is not in the document]
Comment (mohsaka) [reply velox/functions/prestosql/types/IPPrefixType.cpp]: Changed to fmt::format. Is this what you meant? Please let me know. Thanks!

File: velox/functions/prestosql/types/IPPrefixType.cpp [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+      } else {
+        result.append(
+            v6Addr.str() + "/" +
+            std::to_string((uint8_t)intAddr.data()[kIPAddressBytes]));
+      }
+      result.finalize();
+    });
+  }
+
+  static auto splitIpSlashCidr(folly::StringPiece ipSlashCidr) {
```
[the remark this answers is not in the document]
Comment (mohsaka) [reply velox/functions/prestosql/types/IPPrefixType.cpp:122]: I can switch to string view, but the function will get a little messy in that it will be switched to a find + creation of a vector<string_view> + creation of these string_views. This is used for error reporting and is taken straight from

https://github.com/facebook/folly/blob/556e17ac7b1f6cda68de789d5b06e79b2d3b990f/folly/IPAddress.cpp#L75

Please let me know. Thanks!

File: velox/functions/prestosql/types/IPPrefixType.cpp [hunk 15]
```diff
@@ -0,0 +1,329 @@
...
+
+      // Folly allows for creation of networks without a "/"
+      // Check to make sure that we have one
+      if (ipAddressString.str().find('/') == std::string::npos) {
+        context.setStatus(
+            row,
+            threadSkipErrorDetails()
+                ? Status::UserError()
+                : Status::UserError(
+                      "Invalid ipSlashCidr specified. Expected IP/CIDR format, got '{}'",
```
[the remark this answers is not in the document]
Comment (mohsaka) [reply velox/functions/prestosql/types/IPPrefixType.cpp:147]: Error message was taken straight from 
https://github.com/facebook/folly/blob/556e17ac7b1f6cda68de789d5b06e79b2d3b990f/folly/IPAddress.cpp#L101

Changed error message to 
                      "Invalid CIDR IP address specified. Expected IP/PREFIX format, got '{}'",

File: velox/functions/prestosql/types/IPPrefixType.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (mohsaka) [reply velox/functions/prestosql/types/IPPrefixType.cpp]: Update.
Maybe misinterpreted the comment. Changed function argument to std::string_view. Function itself is still using StringPiece.

File: velox/functions/prestosql/types/IPPrefixType.cpp [hunk 16]
```diff
@@ -0,0 +1,328 @@
...
+              context.setStatus(row, Status::UserError());
+              break;
+          }
+        }
+        return;
+      }
+
+      auto net = maybeNet.value();
+      if (net.first.isIPv4Mapped() || net.first.isV4()) {
+        // Take care of mask() throw condition
```
Comment (czentgr) [inline velox/functions/prestosql/types/IPPrefixType.cpp:216]: Nit: do we need this comment? It seems apparent you are handling errors.

**Review by czentgr — COMMENTED** [review]
Looks good. Only some minor comments.

**Comment by czentgr** [comment]
@mohsaka looks like one of the functions tests is failing. Might be related to your new type. Please take a look.

File: velox/functions/prestosql/types/IPPrefixType.cpp [hunk 17]
```diff
@@ -0,0 +1,325 @@
...
+            (uint8_t)intAddr.data()[kIPAddressBytes]));
+      } else {
+        result.append(fmt::format(
+            "{}/{}", v6Addr.str(), (uint8_t)intAddr.data()[kIPAddressBytes]));
+      }
+      result.finalize();
+    });
+  }
+
+  static auto splitIpSlashCidr(const folly::StringPiece& ipSlashCidr) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/types/IPPrefixType.cpp:122]: Might be better to explicitly state the return type of the function.

**Review by aditi-pandit — COMMENTED** [review]
@mohsaka : Did a very high level review. The code looks good. Will read it closely again.

File: velox/functions/prestosql/types/IPPrefixType.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (mohsaka) [reply velox/functions/prestosql/types/IPPrefixType.cpp]: Reverted back to StringPiece, however switched to const &. string_view caused string output issue.
```
Expected error message to contain 'Invalid IP address '2001:db8::1::1'', but received 'Invalid IP address 'ck\0\0\0\0\0\0_����''.
```

File: velox/functions/prestosql/types/IPAddressType.cpp [hunk 18]
```diff
@@ -206,6 +214,57 @@ class IPAddressCastOperator : public exec::CastOperator {
...
+    auto* flatResult = result.as<FlatVector<int128_t>>();
+    const auto* prefixes = input.as<SimpleVector<StringView>>();
+
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      const auto prefix = prefixes->valueAt(row);
+      int128_t addrResult = 0;
+      folly::ByteArray16 addrBytes;
+
+      memcpy(&addrBytes, prefix.data(), kIPAddressBytes);
+      std::reverse(addrBytes.begin(), addrBytes.end());
```
[the remark this answers is not in the document]
Comment (mohsaka) [reply velox/functions/prestosql/types/IPAddressType.cpp:232]: Documented in the docs. 
https://facebookincubator.github.io/velox/develop/types.html#presto-types

**Comment by czentgr** [comment]
@mohsaka Please squash your commits. Otherwise, looks good.

File: velox/docs/functions/presto/conversion.rst [hunk 19]
```diff
@@ -667,52 +700,79 @@ is the number of whole days in the interval, HH is then number of hours between
...
 
   SELECT cast(ipaddress '::ffff:ffff:ffff' as varchar); -- '255.255.255.255'
 
+From IPPREFIX
+^^^^^^^^^^^^^
+
+Casting from IPPREFIX to VARCHAR returns a string formatted as *x.x.x.x/<prefix-length>* for IPv4 formatted IPv6 addresses.
+
+For all other IPv6 addresses it will be formatted in compressed alternate form IPv6 defined in `RFC 4291#section-2.2 <https://datatracker.ietf.org/doc/html/rfc4291.html#section-2.2>`_
+followed by */<prefix-length>*. `[RFC 4291#section-2.3] <https://datatracker.ietf.org/doc/html/rfc4291.html#section-2.3>`_
```
Comment (czentgr) [inline velox/docs/functions/presto/conversion.rst:731]: Nit: does this make `[]` part of the link test? We probably don't want that. Later there is a similar occurrence for `()`.

**Review by czentgr — COMMENTED** [review]
Minor question on the doc external links.

File: velox/docs/functions/presto/conversion.rst (hunk 19 above)
Comment (mohsaka) [reply velox/docs/functions/presto/conversion.rst:731]: Yeah the reason I did that one in brackets is to represent where the source is for the whole sentence is. The one before that is part of the sentence itself so it doesn’t have brackets. I can change it however is needed though

Comment (mohsaka) [reply velox/docs/functions/presto/conversion.rst:731]: Fixed. Standardized. No parenthesis or brackets within sentences. If referring to whole sentences, added with brackets after sentence completion with brackets.

File: velox/docs/functions/presto/conversion.rst [hunk 20]
```diff
@@ -1128,6 +1190,67 @@ Invalid examples:
...
+  SELECT cast('2001:0db8:0000:0000:0000:ff00:0042:8329/129' as ipprefix); -- CIDR value '129' is > network bit count '128'
+  SELECT cast('2001:0db8:0000:0000:0000:ff00:0042:8329/-1' as ipprefix); -- Mask value '-1' not a valid mask
+  SELECT cast('255.2.3.4/33' as ipprefix); -- CIDR value '33' is > network bit count '32'
+  SELECT cast('::ffff:ffff:ffff/33' as ipprefix); -- CIDR value '33' is > network bit count '32'
+
+
+From IPADDRESS
+^^^^^^^^^^^^^^
+
+Returns an IPPREFIX where the prefix length is the length of the entire IP Address.
```
Comment (czentgr) [inline velox/docs/functions/presto/conversion.rst:1244]: Nit: typo "Address".

Comment (mohsaka) [reply velox/docs/functions/presto/conversion.rst]: Fixed.

**Review by czentgr — APPROVED** [review]

**Comment by mohsaka** [comment]
@mbasmanova Could I get a review when you have the time? Thank you!

File: velox/functions/prestosql/types/IPAddressType.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (aditi-pandit) [reply velox/functions/prestosql/types/IPAddressType.cpp:232]: That's fine. Thanks.

File: velox/functions/prestosql/types/IPPrefixType.cpp [hunk 21]
```diff
@@ -0,0 +1,326 @@
...
+  }
+
+  static folly::small_vector<folly::StringPiece, 2> splitIpSlashCidr(
+      const folly::StringPiece& ipSlashCidr) {
+    folly::small_vector<folly::StringPiece, 2> vec;
+    folly::split('/', ipSlashCidr, vec);
+    return vec;
+  }
+
+  static void castFromString(
```
Comment (aditi-pandit) [inline velox/functions/prestosql/types/IPPrefixType.cpp:129]: Is this logic consistent with Java ? Would the same invalid values return the same errors ?

File: velox/functions/prestosql/types/IPPrefixType.cpp [hunk 22]
```diff
@@ -0,0 +1,326 @@
...
+      const SelectivityVector& rows,
+      BaseVector& result) {
+    auto* flatResult = result.as<FlatVector<StringView>>();
+    const auto* ipAddresses = input.as<SimpleVector<int128_t>>();
+
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      auto ipAddress = ipAddresses->valueAt(row);
+      folly::ByteArray16 addrBytes;
+
+      exec::StringWriter<false> result(flatResult, row);
```
Comment (aditi-pandit) [inline velox/functions/prestosql/types/IPPrefixType.cpp:289]: Rename variable to something other than result as its used in the outer scope.

File: velox/functions/prestosql/types/UuidType.cpp [hunk 23]
```diff
@@ -45,7 +45,7 @@ class UuidCastOperator : public exec::CastOperator {
       castFromString(input, context, rows, *result);
     } else {
       VELOX_UNSUPPORTED(
-          "Cast from {} to UUID not yet supported", resultType->toString());
+          "Cast from {} to UUID not yet supported", input.type()->toString());
```
Comment (aditi-pandit) [inline velox/functions/prestosql/types/UuidType.cpp:48]: Is this related to this change ? If not, lets move it to another PR.

File: velox/functions/prestosql/types/IPAddressType.cpp [hunk 24]
```diff
@@ -206,6 +214,57 @@ class IPAddressCastOperator : public exec::CastOperator {
...
+      const SelectivityVector& rows,
+      BaseVector& result) {
+    auto* flatResult = result.as<FlatVector<StringView>>();
+    const auto* ipAddresses = input.as<SimpleVector<int128_t>>();
+
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      auto ipAddress = ipAddresses->valueAt(row);
+      folly::ByteArray16 addrBytes;
+
+      exec::StringWriter<false> result(flatResult, row);
```
[the remark this answers is not in the document]
Comment (mohsaka) [reply velox/functions/prestosql/types/IPAddressType.cpp:251]: Changed to resultWriter.

File: velox/functions/prestosql/types/IPPrefixType.cpp (hunk 21 above)
Comment (mohsaka) [reply velox/functions/prestosql/types/IPPrefixType.cpp:129]: Not the same error message, but will return an error for the same input as java.

File: velox/functions/prestosql/types/IPPrefixType.cpp (hunk 22 above)
Comment (mohsaka) [reply velox/functions/prestosql/types/IPPrefixType.cpp:289]: Same as above.

File: velox/functions/prestosql/types/UuidType.cpp (hunk 23 above)
Comment (mohsaka) [reply velox/functions/prestosql/types/UuidType.cpp:48]: Not related. Sure will open another PR.

Comment (mohsaka) [reply velox/functions/prestosql/types/UuidType.cpp]: Opened here
https://github.com/facebookincubator/velox/pull/11105

File: velox/functions/prestosql/types/IPPrefixType.cpp (hunk 21 above)
Comment (aditi-pandit) [reply velox/functions/prestosql/types/IPPrefixType.cpp:129]: I agree that it might not be possible to use the same errors in all places. But we should be as consistent as we can else its very hard during any kind of traffic migration to validate the builds. Meta has big issues with it.

In general, I feel it might be good to split this work as follows:
i) First PR just adds all the type classes for the IPPrefixType. So this PR minus all the casting code.
ii) Add the fuzzer for IPPrefix Casts that compares with Presto.
iii) Add the cast code with the fuzzer in place. That will ensure that we have most consistency. This would also make it simpler to validate the code you have in IPPrefixCastTest.

Based on prior experience I feel the fuzzer should not be an afterthought.

@amitkdutta @kagamiori @kgpai

**Comment by mohsaka** [comment]
Current split PR for IPPREFIX type only here:
https://github.com/facebookincubator/velox/pull/11122

File: velox/functions/prestosql/types/IPPrefixType.cpp (hunk 21 above)
Comment (mohsaka) [reply velox/functions/prestosql/types/IPPrefixType.cpp:129]: i) https://github.com/facebookincubator/velox/pull/11122
ii) https://github.com/facebookincubator/velox/pull/11209
